src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs

```diff
@@ -2203,5 +2203,99 @@ private int getExactClasses(CORINFO_CLASS_STRUCT_* baseType, int maxExactClasses
             Debug.Assert(index <= maxExactClasses);
             return index;
         }
+
+        private bool getReadonlyStaticFieldValue(CORINFO_FIELD_STRUCT_* fieldHandle, byte* buffer, int bufferSize)
+        {
+            Debug.Assert(fieldHandle != null);
+            Debug.Assert(buffer != null);
+            Debug.Assert(bufferSize > 0);
+
+            FieldDesc field = HandleToObject(fieldHandle);
+            if (field.IsStatic && !field.IsThreadStatic && field.IsInitOnly && field.OwningType is MetadataType owningType)
+            {
+                bool supportedType = field.FieldType.Category switch
+                {
+                    TypeFlags.Enum or
+                    TypeFlags.Class or
+                    TypeFlags.Interface or
+                    TypeFlags.Array or
+                    TypeFlags.SzArray or
+                    TypeFlags.Pointer or
+                    TypeFlags.FunctionPointer => true,
+
+                    _ => field.FieldType.IsPrimitive
+                };
+
+                PreinitializationManager preinitManager = _compilation.NodeFactory.PreinitializationManager;
+                if (supportedType && preinitManager.IsPreinitialized(owningType))
+                {
+                    TypePreinit.ISerializableValue value = preinitManager
+                        .GetPreinitializationInfo(owningType).GetFieldValue(field);
+
+                    int targetPtrSize = _compilation.TypeSystemContext.Target.PointerSize;
+
+                    if (value == null)
+                    {
+                        Debug.Assert(bufferSize >= targetPtrSize);
+
+                        // Write "null" to buffer
+                        new Span<byte>(buffer, targetPtrSize).Clear();
+                        return true;
+                    }
+
+                    if (value.GetRawData(_compilation.NodeFactory, out object data))
+                    {
+                        switch (data)
+                        {
+                            case byte[] bytes:
+                                Debug.Assert(bufferSize >= bytes.Length);
```

Comment:
```suggestion
                                Debug.Assert(bufferSize == bytes.Length);
```
Should the buffer size passed in by the JIT be exact?

Reply:
Yes, can be exact here. Initially I planned it as "get struct's value if it fits into 16 bytes" etc, but yeah, I agree, it should be exact here.